File: ros/ros_packages/rosutility/rosutility/threading/response_waiter.py
import threading

class ResponseWaiter:

    def __init__(self):
        self._cv = threading.Condition()
        self._response = None
    
    def wait_for_response(self):
        with self._cv:
            self._cv.wait()
    
    def resume(self, response):
        self._response = response
        with self._cv:
            self._cv.notify()
    
    def get_response(self):
        return self._response

File: unity/Assets/Scripts/unityagent/Util/Threading/ResponseWaiter.cs
using System.Threading;
using System.Threading.Tasks;

using UnityAgent.SystemMessage;

namespace UnityAgent.Util.Threading
{
public class ResponseWaiter
{
    private readonly int maxWaitTime = 10000; // Milliseconds
    private CancellationTokenSource tokenSource = new CancellationTokenSource();

    public ResponseWaiter()
    {
    }

    public object Result { get; private set; }

    public async Task<object> Wait(int milliseconds=5000)
    {
        int waitTime = WaitTime(milliseconds);
        try
        {
            await Task.Delay(waitTime, tokenSource.Token);
        }
        catch (TaskCanceledException)
        {
        }
        return Result;
    }

    public void Resume(object result)
    {
        Result = result;
        tokenSource.Cancel();
    }

    private int WaitTime(int milliseconds)
    {
        if (IsInAllowedTimeWindow(milliseconds))
            return milliseconds;
        return maxWaitTime;
    }

    private bool IsInAllowedTimeWindow(int milliseconds)
    {
        return !(milliseconds < 0 || milliseconds > maxWaitTime);
    }
}
}